File: tools/conformance/yaml_roundtrip.py
"""Round-trip every yaml-test-suite document this parser accepts.

    parse(input) -> write() -> parse(written)

and ask whether the two agree, twice over:

    by value   the same JSON.  A failure here is data lost or corrupted.
    by event   the same composed event stream.  Strictly harder: it also
               asks that anchors, tags and the text of every scalar survive.

The gap between the two is the writer choosing its own spelling, which is
documented and deliberate - a parse-write cycle is semantically faithful, not
textually faithful.  The first number is the one that has to be 100%.

Nothing in yaml-test-suite tests a writer; the suite is a corpus of inputs.
Running it backwards through this library costs nothing and measures the half
of the module the suite cannot see.

Usage:  yaml_roundtrip.py <suite-dir> <yts-runner> <yts-roundtrip> [-v N]
"""
import collections
import glob
import os
import subprocess
import sys

import yaml

SUITE, RUNNER, ROUNDTRIP = sys.argv[1], sys.argv[2], sys.argv[3]
VERBOSE = 0
if '-v' in sys.argv:
    i = sys.argv.index('-v')
    VERBOSE = int(sys.argv[i + 1]) if len(sys.argv) > i + 1 else 5


def unescape(s):
    """The suite writes invisible characters as glyphs."""
    return (s.replace('␣', ' ')
             .replace('———»', '\t').replace('——»', '\t')
             .replace('—»', '\t').replace('»', '\t')
             .replace('∎', '').replace('←', '\r').replace('⇔', '﻿'))


def run(cmd, src):
    out = subprocess.run(cmd, input=src, capture_output=True)
    return out.stdout.decode('utf-8', 'replace'), out.returncode


by_value = collections.Counter()
by_event = collections.Counter()
failures = []

for path in sorted(glob.glob(SUITE + '/src/*.yaml')):
    tid = os.path.basename(path)[:-5]
    for idx, case in enumerate(yaml.safe_load(open(path))):
        raw = case.get('yaml')
        if raw is None:
            continue
        label = tid if idx == 0 else '%s/%d' % (tid, idx)
        name = case.get('name', '')
        src = unescape(raw).encode('utf-8')

        before, _ = run([RUNNER], src)
        if before.startswith('FAIL'):
            # Refused by design; there is nothing to write back out.
            by_value['skip-refused'] += 1
            by_event['skip-refused'] += 1
            continue

        written, rc = run([ROUNDTRIP, '-w'], src)
        if rc != 0 or written.startswith('FAIL'):
            by_value['write-failed'] += 1
            failures.append((label, name, 'the writer refused it',
                             written.strip()[:70], ''))
        else:
            after, _ = run([RUNNER], written.encode('utf-8'))
            if after.startswith('FAIL'):
                by_value['reparse-failed'] += 1
                failures.append((label, name, 'the output does not parse',
                                 written.strip()[:70], after.strip()[:70]))
            elif after.strip() != before.strip():
                by_value['value-changed'] += 1
                failures.append((label, name, 'the value changed',
                                 before.strip()[:70], after.strip()[:70]))
            else:
                by_value['ok'] += 1

        events, _ = run([ROUNDTRIP], src)
        head = events.split('\n', 1)[0]
        by_event['ok' if head == 'OK' else head.split(':')[0]] += 1

print('=== round trip: parse -> write -> parse, over yaml-test-suite ===')
for title, counter in (('by value (was any data lost?)', by_value),
                       ('by event (did the spelling survive too?)', by_event)):
    attempted = sum(counter.values()) - counter['skip-refused']
    print('\n  %s' % title)
    for k in sorted(counter):
        if k != 'skip-refused':
            print('    %-16s %d' % (k, counter[k]))
    pct = 100.0 * counter['ok'] / attempted if attempted else 0.0
    print('    -> %d of %d documents (%.1f%%)' % (counter['ok'], attempted, pct))

print('\n%d of the suite\'s documents are refused by design and not attempted.'
      % by_value['skip-refused'])

if failures:
    print('\n=== documents whose value did not survive ===')
    for lab, nm, why, a, b in failures:
        print('  %-10s %-36s %s' % (lab, nm[:36], why))
        if VERBOSE:
            if a:
                print('               %s' % a)
            if b:
                print('               %s' % b)
            VERBOSE -= 1

floor = os.environ.get('YTS_RT_MIN')
if floor:
    attempted = sum(by_value.values()) - by_value['skip-refused']
    pct = 100.0 * by_value['ok'] / attempted if attempted else 0.0
    if pct + 0.05 < float(floor):
        print('round trip: %.1f%% by value is below the floor of %s%%'
              % (pct, floor))
        sys.exit(1)
    print('round trip: %.1f%% by value meets the floor of %s%%' % (pct, floor))

File: tools/conformance/yaml_roundtrip.c
/* Read a YAML stream on stdin; write it back out with this library's writer,
   parse that, and report whether the two agree.
 *
 * Prints one of:
 *   OK                 the document survived the trip
 *   SKIP <why>         the input was not a document to begin with
 *   FAIL <stage> <why> the writer or the re-parse disagreed
 *
 * The comparison is over the composed event stream with scalar style left
 * out, because the writer deliberately chooses its own styles (a parse-write
 * cycle is semantically faithful, not textually faithful). Everything else -
 * structure, order, anchors, tags, the text of every scalar - has to match.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ghoti.io/text/yaml.h>

typedef struct {
  char *buf;
  size_t len;
  size_t cap;
  int failed;
} Sink;

static void sink_add(Sink *s, const char *text) {
  size_t n = strlen(text);
  if (s->len + n + 1 > s->cap) {
    size_t cap = s->cap ? s->cap * 2 : 4096;
    while (cap < s->len + n + 1) cap *= 2;
    char *next = realloc(s->buf, cap);
    if (!next) { s->failed = 1; return; }
    s->buf = next;
    s->cap = cap;
  }
  memcpy(s->buf + s->len, text, n);
  s->len += n;
  s->buf[s->len] = '\0';
}

static GTEXT_YAML_Status record(const GTEXT_YAML_Node_Event *ev, void *user) {
  Sink *s = (Sink *)user;
  switch (ev->type) {
    case GTEXT_YAML_NODE_EVENT_STREAM_START: sink_add(s, "+STR\n"); break;
    case GTEXT_YAML_NODE_EVENT_STREAM_END:   sink_add(s, "-STR\n"); break;
    case GTEXT_YAML_NODE_EVENT_DOCUMENT_START: sink_add(s, "+DOC\n"); break;
    case GTEXT_YAML_NODE_EVENT_DOCUMENT_END:   sink_add(s, "-DOC\n"); break;
    case GTEXT_YAML_NODE_EVENT_SEQUENCE_END:   sink_add(s, "-SEQ\n"); break;
    case GTEXT_YAML_NODE_EVENT_MAPPING_END:    sink_add(s, "-MAP\n"); break;
    case GTEXT_YAML_NODE_EVENT_SEQUENCE_START:
    case GTEXT_YAML_NODE_EVENT_MAPPING_START:
      sink_add(s, ev->type == GTEXT_YAML_NODE_EVENT_SEQUENCE_START
          ? "+SEQ" : "+MAP");
      if (ev->anchor) { sink_add(s, " &"); sink_add(s, ev->anchor); }
      if (ev->tag) { sink_add(s, " <"); sink_add(s, ev->tag); sink_add(s, ">"); }
      sink_add(s, "\n");
      break;
    case GTEXT_YAML_NODE_EVENT_SCALAR: {
      sink_add(s, "=VAL");
      if (ev->anchor) { sink_add(s, " &"); sink_add(s, ev->anchor); }
      if (ev->tag) { sink_add(s, " <"); sink_add(s, ev->tag); sink_add(s, ">"); }
      sink_add(s, " ");
      char *copy = malloc(ev->value_len + 1);
      if (!copy) { s->failed = 1; return GTEXT_YAML_E_OOM; }
      memcpy(copy, ev->value, ev->value_len);
      copy[ev->value_len] = '\0';
      sink_add(s, copy);
      free(copy);
      sink_add(s, "\n");
      break;
    }
    case GTEXT_YAML_NODE_EVENT_ALIAS:
      sink_add(s, "=ALI *");
      sink_add(s, ev->value ? ev->value : "");
      sink_add(s, "\n");
      break;
  }
  return s->failed ? GTEXT_YAML_E_OOM : GTEXT_YAML_OK;
}

/* Render a whole parsed stream as events, or NULL. */
static char *events_of(GTEXT_YAML_Document **docs, size_t count) {
  Sink s;
  memset(&s, 0, sizeof(s));
  if (gtext_yaml_stream_walk(docs, count, record, &s) != GTEXT_YAML_OK) {
    free(s.buf);
    return NULL;
  }
  return s.buf ? s.buf : strdup("");
}

int main(int argc, char **argv) {
  /* "-w" prints the written YAML and stops, so a caller can compare the two
     sides any way it likes - by value as well as by event. */
  const int write_only = (argc > 1 && strcmp(argv[1], "-w") == 0);
  static char in[1 << 20];
  size_t n = fread(in, 1, sizeof(in) - 1, stdin);
  in[n] = 0;

  GTEXT_YAML_Error err;
  memset(&err, 0, sizeof(err));
  GTEXT_YAML_Parse_Options popts = gtext_yaml_parse_options_default();
  popts.dupkeys = GTEXT_YAML_DUPKEY_KEEP_ALL;

  size_t count = 0;
  GTEXT_YAML_Document **docs = gtext_yaml_parse_all(in, n, &count, &popts, &err);
  if (!docs) {
    printf("SKIP not a document: %s\n", err.message ? err.message : "?");
    return 0;
  }

  char *before = write_only ? strdup("") : events_of(docs, count);
  if (!before) { printf("FAIL walk before\n"); return 0; }

  /* Write every document back out, one stream. */
  Sink text;
  memset(&text, 0, sizeof(text));
  int wrote = 1;
  {
    GTEXT_YAML_Sink sink;
    if (gtext_yaml_sink_buffer(&sink) != GTEXT_YAML_OK) {
      printf("FAIL sink\n");
      return 0;
    }
    GTEXT_YAML_Write_Options wopts = gtext_yaml_write_options_default();
    /* YTS_RT_BLOCK asks for the block style a human would write, so the two
       defaults can be told apart from the writer itself. */
    if (getenv("YTS_RT_BLOCK")) {
      wopts.pretty = true;
      wopts.flow_style = GTEXT_YAML_FLOW_STYLE_BLOCK;
    }
    GTEXT_YAML_Status wst = (count == 1)
      ? gtext_yaml_write_document(docs[0], &sink, &wopts)
      : gtext_yaml_write_documents(docs, count, &sink, &wopts);
    if (wst != GTEXT_YAML_OK) {
      printf("FAIL write: status %d\n", (int)wst);
      gtext_yaml_sink_buffer_free(&sink);
      wrote = 0;
    } else {
      const char *data = gtext_yaml_sink_buffer_data(&sink);
      size_t len = gtext_yaml_sink_buffer_size(&sink);
      char *copy = malloc(len + 1);
      if (copy) {
        memcpy(copy, data, len);
        copy[len] = 0;
        sink_add(&text, copy);
        free(copy);
      }
      gtext_yaml_sink_buffer_free(&sink);
    }
  }
  if (!wrote) { free(before); free(text.buf); return 0; }

  if (write_only) {
    fwrite(text.buf ? text.buf : "", 1, text.len, stdout);
    if (text.len && text.buf[text.len - 1] != '\n') putchar('\n');
    return 0;
  }

  memset(&err, 0, sizeof(err));
  size_t count2 = 0;
  GTEXT_YAML_Document **docs2 =
      gtext_yaml_parse_all(text.buf ? text.buf : "", text.len, &count2, &popts, &err);
  if (!docs2) {
    printf("FAIL reparse: %s\n", err.message ? err.message : "?");
    printf("---- written ----\n%s----\n", text.buf ? text.buf : "");
    return 0;
  }

  char *after = events_of(docs2, count2);
  if (!after) { printf("FAIL walk after\n"); return 0; }

  if (strcmp(before, after) != 0) {
    printf("FAIL differs\n");
    printf("---- written ----\n%s", text.buf ? text.buf : "");
    printf("---- before ----\n%s---- after ----\n%s", before, after);
  } else {
    printf("OK\n");
  }

  free(before); free(after); free(text.buf);
  for (size_t i = 0; i < count; i++) gtext_yaml_free(docs[i]);
  free(docs);
  for (size_t i = 0; i < count2; i++) gtext_yaml_free(docs2[i]);
  free(docs2);
  return 0;
}
